pkg/openstack/routes.go

```diff
@@ -59,12 +67,93 @@ func NewRoutes(os *OpenStack, network *gophercloud.ServiceClient, atomicRoutes b
 		return nil, errors.ErrNoRouterID
 	}
 
-	return &Routes{
+	routes := Routes{
 		network:             network,
 		os:                  os,
 		atomicRoutes:        atomicRoutes,
 		allowedAddressPairs: allowedAddressPairs,
-	}, nil
+	}
+
+	err := routes.getOrCreateNodeSecurityGroup(os.routeOpts.RouterID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &routes, nil
+}
+
+func (r *Routes) getOrCreateNodeSecurityGroup(router_id string) error {
+	// TODO(jeffyjf) Maybe one router can be used to support multiple
+	// clusters. Whether it is appropriate that use router id as the security
+	// group name's identity. Maybe we should introduce other mechanism to identify
+	// the security group.
+	sgName := fmt.Sprintf("k8s-node-sg-%s", router_id)
+	sgId, err := secgroups.IDFromName(r.network, sgName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			mc := metrics.NewMetricContext("security_group", "create")
+			group, err := groups.Create(r.network, groups.CreateOpts{Name: sgName}).Extract()
+			if mc.ObserveRequest(err) != nil {
+				return err
+			}
+			sgId = group.ID
+		} else {
+			return err
+		}
+	}
+	r.nodeSecurityGroupId = sgId
+	return nil
+}
+
+// checkSecurityGroupRules check whether the ingress security group rules that related to the route is existing.
+// these security group rules use to ensure other nodes permit the traffic from this node (nodeName) pass through.
+func (r *Routes) checkSecurityGroupRules(route routers.Route, existingRules []rules.SecGroupRule, nodeName types.NodeName, nodes []*v1.Node) bool {
+	sgRuleForPodCidrFound := false
+	sgRuleForNodeAddressFound := false
+	ip, _, _ := net.ParseCIDR(route.DestinationCIDR)
+	isCIDRv6 := ip.To4() == nil
+	nodeAddr := getAddrByNodeName(nodeName, isCIDRv6, nodes)
+	nodeAddrCidr := fmt.Sprintf("%s/32", nodeAddr)
+
+	for _, rule := range existingRules {
+		// route.DestinationCIDR is equivalent with the node (nodeName)'s PodCidr
+		if route.DestinationCIDR == rule.RemoteIPPrefix {
+			sgRuleForPodCidrFound = true
+		}
```

Comment:
What if a rule opening broader traffic is there already? Should we just check if `DestinationCIDR` is in `RemoteIPPrefix`, not is equal?